[addons][vfs] cleanup, "C" interface fix, documentation rework

AlwinEsch opened the pull request:
## Description

This request revises the VFS addon interface. The changes in the addon itself are rather minor, since only the `CVFSEntry::IoControl` has been revised on the addon to avoid the Kodi-side header "IFileTypes.h" (this only corresponded to C++).

This API change now translates the Kodi site to its own for addon, which makes this "C" ABI compatible and more importantly, safer in the event of changes to IFileTypes.h, which could inadvertently cause problems with addons.

Another API change is that the individual structures are now in their own arrays, so it is safer if a callback function or property is added. However, this does not require any code changes to addon and is fixed with a new creation.

The commit **"[settings] allow TranslatableIntegerSettingOption to get string from"** allows the addon itself a string ID to its network drive connection that can be created in Kodi's settings. This is independent and will still be moved in a own request.

There are also some clang code cleanups in it.

In addition, everything between Kodi and addon is now in pure "C".

Here is the new documention view available: http://alwinesch.github.io/group__cpp__kodi__addon__vfs.html

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->

## Motivation and Context
Related to https://github.com/xbmc/xbmc/issues/17514 and to become the documentation in future available for everything :grin: 

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
Tested now on few addons, need to test more and comes next days, also is the API not so big that it becomes critical.

<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch committed:
[addons][vfs] clang style cleanup on addon header

AlwinEsch committed:
[addons][vfs] do not interact addon class with Kodi

This was before given to Kodi to have on his calls to addon.
That basically not match a "C" ABI where the use of "C++" parts within
"C" not allowed!

With this it becomes given via "void*" to Kodi and with a "static_cast"
changed back to required class on addon.

AlwinEsch committed:
[addons][vfs] add "#ifdef __cplusplus" to addon header

Added to separate "C" and "C++" part and further to allow a "C" only
in future.

AlwinEsch committed:
[addons][vfs] make all "C" interface structures to own memory

This done to prevent API backward problems if something becomes replaced
e.g. on props or to kodi calls.

AlwinEsch committed:
[addons][vfs] change ABI interface to match pure "C"

This changes the interface to pure "C" and no "C++" dependencies in it.

AlwinEsch committed:
[addons][vfs] increase API version to 2.3.0

garbear commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/VFS.h:
```diff
@@ -135,9 +138,9 @@ extern "C"
-    KodiToAddonFuncTable_VFSEntry toAddon;
+    AddonProps_VFSEntry* props;
+    AddonToKodiFuncTable_VFSEntry* toKodi;
+    KodiToAddonFuncTable_VFSEntry* toAddon;
```
what about std::unique ptrs?

AlwinEsch replied:
We can not use a C++ `std::unique` in a "C" ABI, thats one of the main reason of this request.

garbear replied:
oh yea, didn't look at the location of this file.

AlwinEsch replied:
NP, thanks a lot for review!

garbear commented on xbmc/network/GUIDialogNetworkSetup.h:
```diff
@@ -24,6 +24,7 @@ class CGUIDialogNetworkSetup : public CGUIDialogSettingsManualBase
     int defaultPort;       //!< Default port to use for protocol
     std::string type;      //!< URL type for protocol
     int label;             //!< String ID to use as label in dialog
+    std::string addonId; //!< Addon identifier, leaved empty if inside Kodi
```
It's not evident to me, where is this used?

AlwinEsch replied:
Currently vfs addons are forced to take Kodi's own string id's (e.g. at `vfs.sftp` https://github.com/xbmc/vfs.sftp/blob/Leia/vfs.sftp/addon.xml.in#L21) and the `std::pair<int, int>` in settings this could not be identified.

This is then taken here https://github.com/xbmc/xbmc/pull/17527/files#diff-8211fa801621c92df9bf6515d2d45b8aR200 to give to replaced `pair` with `TranslatableIntegerSettingOption` his addon id.

garbear replied:
thanks for the explanation

garbear commented:
Two comments, otherwise looks good

AlwinEsch committed:
[addons][vfs] rework documentation

With this change the whole VFS documentation becomes reworked and to have
usable on doxygen.

There are much parts added to understand his usage.

AlwinEsch commented on xbmc/addons/AddonBindings.cmake:
```diff
@@ -5,7 +5,6 @@
 # Keep this in alphabetical order
 set(CORE_ADDON_BINDINGS_FILES
     ${CORE_SOURCE_DIR}/xbmc/cores/AudioEngine/Utils/AEChannelData.h
-    ${CORE_SOURCE_DIR}/xbmc/filesystem/IFileTypes.h
```
As note, to remove here was a bad idea from me. This header also be used currently by PVR on `libXBMC_addon.h`.

This line change becomes removed.

AlwinEsch committed:
[addons][vfs] prevent C++ designed Kodi Header IFileTypes.h

On this commit becomes them replaced to use independent enum and
structures, to pass related data between Kodi and Addon.

This prevents conflicts when using "C" and ensures a secure ABI interface.

In addition, it is much safer in the event that the Kodi changes what in its
headers. This can easily occur because previously it was not directly visible
that IFileTypes.h was also used on addons.

Only currently still used on libXBMC_addon.h and removed on PVR rework.

lrusak commented:
jenkins build this please

AlwinEsch commented:
If nothing against then I bring it in tommorow and then update of the addons.
Working test of all vfs addons was OK.

jenkins build this please
